### PR TITLE
[2.6] Audit log enhancements for Kubectl access path

### PR DESCRIPTION
**Pull request by prachidamle**
PR for 2.6.x branch from the original PR https://github.com/rancher/rancher/pull/33268

https://github.com/rancher/rancher/issues/33126

This fix is to support audit logging from kubectl access which uses kubeconfig tokens. In case of kubeconfig token, the AuthProvider is always set to "local", hence the backend external userPrincipal is not found and is missing information. Hence the audit log data is not seen.
As a fix, the kubeconfig token is now fixed to carry the correct authProvider going forward.

Another part of the fix is instead reading extraInfo from user's login token, this is now read from userattributes where the extra info is saved after user logs in. This information can be found using the userId extracted from any token.
The userAttributes.Extra is also saved to ClusterUserAttributes.Extra field.

The process of finding the extra attributes for the user will be, for every API request:

Check if extra attributes are found in userattributes for the token's provider using the userId
If the token's provider is empty or is local, then add all extraAttributes into the Audit entry to cover all IDP(external authProvider) values
The fallback if extra attributes are not found, is to read from the token if found, or otherwise read from the user object

### CODE REVIEW

**Review by snasovich — COMMENTED** [review]
Left some comments from general coding perspective.
Definitely need someone more familiar with the inner workings of auth to review.

**Review by brendarearden — COMMENTED** [review]
This PR looks good!

**Review by brendarearden — APPROVED** [review]

File: pkg/auth/tokens/manager.go [hunk 1]
```diff
@@ -618,6 +628,13 @@ func (m *Manager) UserAttributeChanged(attribs *v32.UserAttribute, provider stri
 		}
 	}
 
+	if attribs.ExtraByProvider == nil && extraInfo != nil {
```
[the remark this answers is not in the document]
Comment (prachidamle) [reply pkg/auth/tokens/manager.go:631]: Right, but first we need to check the case when the very first time an extraInfo is being saved to the userAttributes.ExtraByProvider for some provider. So we check if the outer map is nil but extraInfo  is not nil for some provider. 
Later reflect.DeepEqual is used to check the inner map for the provider, obtained from attribs.ExtraByProvider[provider]
`reflect.DeepEqual(attribs.ExtraByProvider[provider], extraInfo)`

**Review by ryansann — DISMISSED** [review]

File: pkg/apis/cluster.cattle.io/v3/types.go [hunk 2]
```diff
@@ -14,11 +14,11 @@ type ClusterUserAttribute struct {
...
-	Groups       []string            `json:"groups,omitempty"`
-	LastRefresh  string              `json:"lastRefresh,omitempty"`
-	NeedsRefresh bool                `json:"needsRefresh"`
-	Enabled      bool                `json:"enabled"`
-	Extra        map[string][]string `json:"extra,omitempty"`
+	Groups          []string                       `json:"groups,omitempty"`
+	LastRefresh     string                         `json:"lastRefresh,omitempty"`
+	NeedsRefresh    bool                           `json:"needsRefresh"`
+	Enabled         bool                           `json:"enabled"`
+	ExtraByProvider map[string]map[string][]string `json:"extraByProvider,omitempty"`
```
Comment (kinarashah) [inline pkg/apis/cluster.cattle.io/v3/types.go:21]: I am not 100% sure where all ClusterUserAttribute is used, but we decided to deprecate `Extra` field. Are there any use cases we should be concerned/handling with this?

File: pkg/api/norman/customization/cluster/shell.go [hunk 3]
```diff
@@ -35,8 +36,16 @@ func (s *ShellLinkHandler) LinkHandler(apiContext *types.APIContext, next types.
...
 		shellTTL = minutes * 60 * 1000 // convert minutes to milliseconds
 	}
-
-	token, err := userManager.EnsureToken("kubectl-shell-"+userID, "Access to kubectl shell in the browser", "kubectl-shell", userID, &shellTTL, true)
+	input := user.TokenInput{
+		TokenName:    "kubectl-shell-" + userID,
+		Description:  "Access to kubectl shell in the browser",
+		Kind:         "kubectl-shell",
+		UserName:     userID,
+		AuthProvider: "local",
```
[the remark this answers is not in the document]
Comment (kinarashah) [reply pkg/api/norman/customization/cluster/shell.go:44]: P.S: Saw we did the same for telemetry token as well, so are we using local for our internal tokens? In that case, could we comment the context + have a local provider constant so it's clear for when new tokens get introduced.

File: pkg/auth/tokens/manager.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (snasovich) [reply pkg/auth/tokens/manager.go:631]: Oh, I've missed `attribs.ExtraByProvider[provider]` vs. `attribs.ExtraByProvider` - my bad.

**Review by snasovich — DISMISSED** [review]
My comments have been addressed and it's LGTM.

File: pkg/api/norman/customization/cluster/shell.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (prachidamle) [reply pkg/api/norman/customization/cluster/shell.go:44]: yes, internal tokens default to "local" is what I see. I will update it to use a constant.

File: pkg/apis/cluster.cattle.io/v3/types.go (hunk 2 above)
Comment (prachidamle) [reply pkg/apis/cluster.cattle.io/v3/types.go:21]: ClusterUserAttribute is used by the ACE endpoint and its webhook. It had the field Extra, but was not set and used until now by ACE codebase https://github.com/rancher/kube-api-auth/blob/ca50db33ff69ec47636ae21f714ab741750d58bf/pkg/service/handlers/v1-kube-api-authn.go

The change in this PR will help us fix this Audit for ACE path as well in future https://github.com/rancher/rancher/issues/33330


Rancher has the [controllers](https://github.com/rancher/rancher/blob/release/v2.6/pkg/controllers/managementuser/clusterauthtoken/user_attribute.go#L26) to keep the data in ClusterUserAttribute in sync with the v3.UserAttribute CR which holds information about the user principals. In this PR I have added the `ExtraByProvider` newly to the v3.UserAttribute - it was not used before.

**Review by kinarashah — APPROVED** [review]

**Pull request merged**